apis/cr/v1/cluster.go

```diff
@@ -97,4 +108,29 @@ const (
 	PgclusterStateInitialized PgclusterState = "pgcluster Initialized"
 	// PgclusterStateRestore ...
 	PgclusterStateRestore PgclusterState = "pgcluster Restoring"
+
+	// PodAntiAffinityRequired results in requiredDuringSchedulingIgnoredDuringExecution for any
+	// default pod anti-affinity rules applied to pg custers
+	PodAntiAffinityRequired PodAntiAffinityType = "required"
+
+	// PodAntiAffinityPreffered results in preferredDuringSchedulingIgnoredDuringExecution for any
+	// default pod anti-affinity rules applied to pg custers
+	PodAntiAffinityPreffered PodAntiAffinityType = "preferred"
+
+	// PodAntiAffinityDisabled disables any default pod anti-affinity rules applied to pg custers
+	PodAntiAffinityDisabled PodAntiAffinityType = "disabled"
 )
+
+// ValidatePodAntiAffinityType is responsible for validating whether or not the type of pod
+// anti-affinity specified is valid
+func (p PodAntiAffinityType) ValidatePodAntiAffinityType() error {
```

Comment:
I will go ahead and change to `Validate()` to align with some other similar validation functions we have, e.g. those in the benchmark API.

As for what is returned (i.e. `error` vs `bool`) I would like to defer any changes and discuss further, especially in regards to how we perform validation across the PostgreSQL Operator in general moving forward.  My problem with simply returning `bool` is that all context for why it failed validation, or if more than one failure occurred (although not applicable in the above use case), is lost.  And since an error here will often result in an error message being logged somewhere in the operator logs, or in the case of the apiserver will be returned to the caller (e.g. as an error response to the `pgo create cluster`command), I want to ensure there is more than simply "invalid value", e.g. context around why it was invalid and what value it instead should/could have been.  

I could require the caller to display the pertinent error information, but now the caller, and not the model where the various types actually live, is responsible for figuring out why it failed and displaying the proper error (and error messages can always be ignored by the caller anyway, allowing them to display any error they wish, if at all).  With this approach I have a clear error message that can easily be identified in the logs, or by any UI end user, in a consistent way, allowing them to easily correct the validation issue (and without having to re-implement and continuously update logic for why something failed everywhere this is called, e.g. all UI layers, the operater itself, apiserver, etc.).

In general (and while not currently implemented in our code), when it comes to designing validation I tend to think of Matrin Fowler's notification pattern, which results in validation error information being returned to the user in the form of collections of errors, which each provide context as to what validation failed and why.  Anyway, considering the above, I'd like to defer any further changes here pending further discussion around our overall strategy and design for PostgreSQL Operator validation.

Reply:
WFM

Reply:
> As for what is returned (i.e. error vs bool) I would like to defer any changes and discuss further, especially in regards to how we perform validation across the PostgreSQL Operator in general moving forward.

I can accept that.

> And since an error here will often result in an error message being logged somewhere in the operator logs, or in the case of the apiserver will be returned to the caller (e.g. as an error response to the pgo create clustercommand), I want to ensure there is more than simply "invalid value", e.g. context around why it was invalid and what value it instead should/could have been.

This is what I mean by mixing 'presentation'. I see value in communicating back what is 'wrong' about the particular thing being validated, however, I feel that `error` is the wrong place to do this. I would obviously agree that an error is meant to convey something/what went wrong. But I believe there are better ways for us to do that than simply `error`. For example, I'd rather see something implement the `Error` interface, that allows for 1) allows it to be handled as an `error` but 2) provides a mechanism to get more information about what went wrong without encoding it into a string.

At any rate, I'm fine with deferring for now. But it is certainly something we need to revisit and address in a more idiomatic way.

Reply:
And FWIW, the new Wrap/Unwrap feature that was introduced in 1.13 would be an acceptable approach. It would enable a reusable type across all validation, while allowing for the composition/collection of errors for what specifically went wrong as a result.

Food for thought.

https://blog.golang.org/go1.13-errors

Reply:
Yea, great feature - prior to 1.13 would reach for the following to get wrap/unwrap functionality: